BalloonTip: Don't hide when the BalloonTip blocks the cursor

Dentomologist opened the pull request:
On master, putting the cursor in the part of a ToolTipWidget where the arrow would be prevents the BalloonTip from showing up (or flickering for a moment and immediately closing). This happens because once the cursor is over the BalloonTip it's no longer in the ToolTipWidget, triggering the latter's leaveEvent which closes the balloon.

In this PR, when that happens the BalloonTip tracks the cursor until it either leaves the ToolTipWidget's bounding box or leaves the BalloonTip and goes back to the ToolTipWidget, then respectively closes the BalloonTip or leaves it open.

iwubcode reviewed (approved):
Code LGTM.  Untested.

Dentomologist commented:
Converting to draft until I address some edge cases with alt tabbing.

While I'm on the subject, on master BalloonTips will appear even when Dolphin doesn't have focus as long as the ToolTipWidget isn't blocked by anything; do we want to keep that behavior or should I change it?

Dentomologist commented:
Ok, time to pick this PR back up.

I decided to keep the current behavior of allowing the tooltip to open even when Dolphin is in the background (as long as the ToolTipWidget is still visible), which matches the behavior of two of the three other applications I tested (Chrome and Visual Studio).

Dentomologist commented:
Rebased and made a few minor changes:

* Added forward declarations for QEvent and related classes.
* QEvents are now passed to their base classes after being handled. I don't think this has any meaningful effect but it's good practice.
* Renamed `ToolTipWidget::KillAndHide()` to `KillTimerAndHideBalloon` for clarity.
